libbeat/autodiscover/common/config.go

```diff
@@ -0,0 +1,63 @@
+package common
+
+import (
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/bus"
+	"github.com/elastic/beats/libbeat/logp"
+	ucfg "github.com/elastic/go-ucfg"
+)
+
+// ConfigMapping maps discover event keys with possible values, if a match happens the given config will be used
+type ConfigMapping map[string]map[string][]*common.Config
+
+// GetConfig returns a matching Config if any, nil otherwise
+func (c ConfigMapping) GetConfig(event bus.Event) []*common.Config {
+	for key, mappings := range c {
+		value, ok := event[key]
+		if !ok {
+			return nil
+		}
+
+		for expected, configs := range mappings {
+			if value == expected {
+				var result []*common.Config
+				// unpack input
+				vars, err := ucfg.NewFrom(map[string]interface{}{
+					"data": event,
+				})
+				if err != nil {
+					logp.Err("Error building config: %v", err)
+				}
+				opts := []ucfg.Option{
+					ucfg.PathSep("."),
+					ucfg.Env(vars),
+					ucfg.ResolveEnv,
+					ucfg.VarExp,
+				}
+				for _, config := range configs {
+					c, err := ucfg.NewFrom(config, opts...)
```

Comment:
Why do you need ucfg here?

Minor problem with this approach is, you are loosing the full setting paths. If some config unpacking fails, the setting it's full path is reported with the error message.

Reply:
I pass provider event data for variable expansion, so config templates can use data from the event. Agree to the minor issue (would affect logs mainly?)

Reply:
Yes, it's for logs only. As we can't store file positions (by line) yet, not having the correct names can make it more difficult to track-down misconfigurations.

Reply:
discovery can be more automatic than providing `docker-image: redis`. if we use docker labels, we should be able to pass `annotation` like information same like collectbeat to be able to discover workloads automatically. i could run 10 containers in a host. it would be a huge config if i have to say for each via config how it has to be discovered.

Reply:
I agree @vjsamuel, this is a first PR, intended to bring the basic functionality and building blocks to grow from there. Definitely that's something we want to have